File: 2022/16/java/Layout.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class Layout {
	// --- Constants ---
	private static final String REGEX = "Valve ([A-Z]{2}) has flow rate=(\\d+); tunnels? leads? to valves? ([A-Z]{2}(,\\s*[A-Z]{2})*)";

	// --- Data ---
	private final Set<String> valves;
	private final Map<String, Long> flows;
	private final Map<String, List<String>> tunnels;
	private final Set<String> usefulValves;
	private Map<String, Map<String, Integer>> all;

	private Layout() {
		valves = new HashSet<>();
		flows = new HashMap<>();
		tunnels = new HashMap<>();
		usefulValves = new HashSet<>();
		all = new HashMap<>();
	}

	// --- Read and parse the input file ---
	public static Layout load(String filename) throws IOException {
		Layout layout = new Layout();

		String text = Files.readString(Paths.get(filename));
		Pattern pattern = Pattern.compile(REGEX);
		Matcher matcher = pattern.matcher(text);
		while (matcher.find()) {
			String id = matcher.group(1);
			long flow = Long.parseLong(matcher.group(2));
			String[] routes = matcher.group(3).split("\\s*,\\s*");
			layout.valves.add(id);
			layout.flows.put(id, flow);
			layout.tunnels.put(id, Arrays.asList(routes));
			if (flow > 0) layout.usefulValves.add(id);
		}
		layout.all = layout.fw();
		return layout;
	}

	// --- Floyd Warshall Algorithm ---
	private Map<String, Map<String, Integer>> fw() {
		Map<String, Map<String, Integer>> distances = new HashMap<>();
		for (String from: valves) {
			Map<String, Integer> map = new HashMap<>();
			distances.put(from, map);
			for (String to: valves) {
				map.put(to, valves.size() + 1);
			}
		}

		for (String from: tunnels.keySet()) {
			for (String to: tunnels.get(from)) {
				distances.get(from).put(to, 1);
			}
		}
		for (String from: valves) {
			distances.get(from).put(from, 0);
		}

		for (String via: valves) {
			for (String from: valves) {
				for (String to: valves) {
					Integer d = distances.get(from).get(via) + distances.get(via).get(to);
					if (distances.get(from).get(to) > d) {
						distances.get(from).put(to,  d);
					}
				}
			}
		}
		return distances;
	}

	private long getFlow(Set<String> set) {
		return set.stream().mapToLong(x -> flows.get(x).longValue()).sum();
	}

	// --- Depth First Search: Puzzle 1 ---
	public long dfs1() {
		Set<String> open = new HashSet<>();
		return dfs1("AA", 0, 0, open);
	}

	private long dfs1(String current, long time, long total, Set<String> open) {
		long max = total + getFlow(open) * (30 - time);
		for (String next: usefulValves) {
			if (open.contains(next)) continue;
			long delta = all.get(current).get(next) + 1;
			if (time + delta >= 30) continue;
			long newTotal = total + delta * getFlow(open);
			open.add(next);
			long value = dfs1(next, time + delta, newTotal, open);
			if (max < value) max = value;
			open.remove(next);
		}
		return max;
	}

	// --- Depth First Search: Puzzle 2 ---
	public long dfs2() {
		Set<String> open = new HashSet<>();
		return dfs2("AA", false, 0, 0, open, usefulValves, 0);
	}

	private long dfs2(String current, boolean elephant, long time, long total, Set<String> open, Set<String> useful, long totalFlow) {
		long max = total + totalFlow * (26 - time);
		if (!elephant) {
			Set<String> newCandidates = new HashSet<>(useful);
			for (String v: open) newCandidates.remove(v);

			Set<String> newOpen = new HashSet<>();
			long maxElephant = dfs2("AA", true, 0, 0, newOpen, newCandidates, 0);
			max = total + totalFlow * (26 - time) + maxElephant;
		}

		for (String next: useful) {
			if (open.contains(next)) continue;
			long delta = all.get(current).get(next) + 1;
			if (time + delta >= 26) continue;
			long newTotal = total + delta * totalFlow;
			open.add(next);
			long nextFlow = flows.get(next).longValue();
			totalFlow += nextFlow;
			long value = dfs2(next, elephant, time + delta, newTotal, open, useful, totalFlow);
			if (max < value) max = value;
			open.remove(next);
			totalFlow -= nextFlow;
		}
		return max;
	}
}

File: 2022/16/javascript/layout.mjs
import fs from "node:fs";

// --- Constants ---
const REGEX = /Valve ([A-Z]{2}) has flow rate=(\d+); tunnels? leads? to valves? ([A-Z]{2}(,\s*[A-Z]{2})*)/g

// --- Creates a new Layout from the input file ---
function createLayout(filename) {
	// --- Variables ---
	let valves = new Set();
	let flows = {};
	let tunnels = {};
	let usefulValves = new Set();

	// --- Reads and parse the input file ---
	const text = fs.readFileSync(filename, "utf8");
	for (const cap of text.matchAll(REGEX)) {
		let id = cap[1];
		let flow = Number(cap[2]);
		let routes = cap[3].split(/\s*,\s*/);
		valves.add(id);
		flows[id] = flow;
		tunnels[id] = routes;
		if (flow > 0) usefulValves.add(id);
	}
	const all = FW(valves, tunnels);

	const getFlow = function(set) {
		let total = 0;
		for (const x of set) total += this.flows[x];
		return total;
	}

	const dfs1 = function() { return DFS1(this, "AA", 0, 0, new Set()); }
	const dfs2 = function() { return DFS2(this, "AA", false, 0, 0, new Set(), this.usefulValves, 0); }

	return { flows, usefulValves, all, getFlow, dfs1, dfs2 };
}

// --- Floyd Warshall Algorithm ---
function FW(valves, tunnels) {
	let distances = {};
	for (const from of valves.values()) {
		let map = {};
		for (const to of valves.values()) {
			map[to] = valves.size + 1;
		}
		distances[from] = map;
	}

	for (const from of Object.keys(tunnels)) {
		for (const to of tunnels[from]) {
			distances[from][to] = 1;
		}
	}

	for (const from of valves.values()) {
		distances[from][from] = 0;
	}

	for (const via of valves.values()) {
		for (const from of valves.values()) {
			for (const to of valves.values()) {
				let d = distances[from][via] + distances[via][to];
				if (distances[from][to] > d) {
					distances[from][to] = d;
				}
			}
		}
	}
	return distances;
}

// --- Depth First Search: Puzzle 1 ---
function DFS1(map, current, time, total, open) {
	let max = total + map.getFlow(open) * (30 - time);
	for (const next of map.usefulValves) {
		if (open.has(next)) continue;
		const delta = map.all[current][next] + 1;
		if (time + delta >= 30) continue;
		let newTotal = total + delta * map.getFlow(open);
		open.add(next);
		const value = DFS1(map, next, time + delta, newTotal, open);
		if (max < value) max = value;
		open.delete(next);
	}
	return max;
}

// --- Depth First Search: Puzzle 2 ---
function DFS2(map, current, elephant, time, total, open, useful, totalFlow) {
	let max = total + totalFlow * (26 - time);
	if (!elephant) {
		let newCandidates = new Set(useful);
		for (const v of open) newCandidates.delete(v);
		let newOpen = new Set();
		let maxElephant = DFS2(map, "AA", true, 0, 0, newOpen, newCandidates, 0);
		max = total + totalFlow * (26 - time) + maxElephant;
	}

	for (const next of useful) {
		if (open.has(next)) continue;
		const delta = map.all[current][next] + 1;
		if (time + delta >= 26) continue;
		const newTotal = total + delta * totalFlow
		open.add(next);
		const nextFlow = map.flows[next];
		totalFlow += nextFlow;
		const value = DFS2(map, next, elephant, time + delta, newTotal, open, useful, totalFlow);
		if (max < value) max = value;
		open.delete(next);
		totalFlow -= nextFlow;
	}
	return max;
}

export default createLayout;
